proposals/0000-exception-backtraces.rst

```diff
@@ -224,6 +345,116 @@ the fact that stack traces are primarily a debugging tool and somewhat of a
+proposal about how to reach two goals
+(<https://github.com/ghc-proposals/ghc-proposals/pull/330>):
+- Keep the migration costs as low as possible. I.e. most usages should work
+  without any change.
+- Get a type checking error when the changed exception structure breaks existing
+  code. (I.e. no silent changes in behaviour.)
+
+The solution presented in this proposal has been agreed upon by all involved
+parties.
+
+``catch``and ``handle`` work with both, ``SomeExceptionWithBacktrace`` and
+``SomeException``. This is the main reason for keeping ``SomeException``
+as a layer in the exception hierarchy.
+
+The usages of ``throw`` and ``throwIO`` don't have to be changed, too.
+In fact, it showed that most submodules of GHC don't need any changes.
+Only a few changes were needed to be made to GHC itself (e.g. in ``compiler/``).
+
+Direct calls to ``toException`` and ``fromException``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+As the types of ``toException`` and ``fromException`` changed, calls to them might
+need to be adjusted to the new types. This is guided by type checking errors.
+
+Usually, when these calls aren't bound to ``SomeException`` specific types, no
+change should be needed.
+
+In expressions where ``fromException``'s changed parameter type leads to type 
+errors, one common solution is to convert the exception value first with
+``toException``.
+
+As an example let's consider we don't want to change the type of 
+``GHC.TopHandler.real_handler`` to keep the needed migration effort small. ::
+
+    real_handler :: (Int -> IO a) -> SomeException -> IO a
+    real_handler exit se = do
+      flushStdHandles -- before any error output
+      -- The call to fromException needs to be preceded by a call to
+      -- toException. (Original line from master commented out.)
+      -- case fromException se of
+      case (fromException . toException) se of
+          Just StackOverflow -> do
+              reportStackOverflow
+              exit 2
+
+          ... [Other cases]
+
+Type synonym
+~~~~~~~~~~~~
+
+As long as ``SomeExceptionWithBacktrace`` isn't deconstructed in pattern
+matches, a simple type synonym may be used to provide backwards
+compatibility. ::
+
+    module Control.Monad.Catch where
+
+    #if __GLASGOW_HASKELL__ < 903
+    type SomeExceptionWithBacktrace = SomeException
+    #endif
+
+    catchAll :: MonadCatch m => m a -> (SomeExceptionWithBacktrace -> m a) -> m a
+
+    handleAll :: MonadCatch m => (SomeExceptionWithBacktrace -> m a) -> m a -> m a
+
+This is only needed when ``SomeExceptionWithBacktrace`` should be used as type
+in the program (e.g. to be able to access the backtraces).
+As already discussed, in very most cases it is fine to stick with
+``SomeException`` which is supported by old versions of GHC and those that
+implement this proposal.
+
+In general, it should always be considered if a "down cast" to ``SomeException``
```

Comment:
Yes, although it would be worthwhile pointing out that such a down-cast would lose backtrace information.

Reply:
@bgamari , sounds reasonable. Will do.

Reply:
Done.